core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -80,114 +146,344 @@
+    * steps, such as connecting via socket or opening file descriptors.
     * <p>
-    * This method is guaranteed to not be invoked concurrently with other operations. This also means another method
+    * The provided {@link InitializationContext} contains many helpful objects, including the configuration of the
+    * cache and store, concurrency utilities such as {@link org.infinispan.util.concurrent.BlockingManager} or
+    * an executor reserved for non blocking operations only {@link InitializationContext#getNonBlockingExecutor()}.
+    * <p>
+    * This method is guaranteed to not be invoked concurrently with other operations. This means another method
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@code start} a second time if
+    * {@link #stop()} has been invoked and allowed for its stage to complete.
     * @param ctx initialization context used to initialize this store
     * @return a stage that when complete signals that this store has been successfully started
     */
    CompletionStage<Void> start(InitializationContext ctx);
 
    /**
+    * This method will be invoked is being shut down. It is expected that all resources related to the store to be
+    * freed upon completion of the returned stage.
     * <p>
     * This method is guaranteed to not be invoked concurrently with other operations. This also means another method
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@link #start(InitializationContext)}
+    * a second time if {@code stop} has been invoked and allowed for its stage to complete.
     * @return a stage that when complete signals that this store has been stopped
     */
    CompletionStage<Void> stop();
 
    /**
-    *
-    * @return
+    * Returns a set of characteristics for this store and its elements. This method will be invoked one to many times
+    * to determine which methods of the store can be used and how its data can be handled.
+    * <p>
+    * Please see {@link Characteristic} and its values for a description of what each characteristic declares the
+    * store as supporting.
+    * @implSpec
+    * The default implementation returns an empty set
+    * @return the set of characteristics that this store supports
     */
    default Set<Characteristic> characteristics() {
       return EnumSet.noneOf(Characteristic.class);
    }
 
    /**
-    *
-    * @return
+    * Returns a stage that when complete will tell if the current store can be accessed for requests. This can be useful
+    * for some store implementations that may rely on an external source that may become unreachable. This can reduce
+    * sending requests to a store that is not available as future Cache requests will all be met with
```

Comment:
"not available, as subsequent cache requests will result in a {@link StoreUnavailableException} being thrown until the store becomes available again."